File: py/zbc/address.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
"""Account addresses: the ZBC_/ZNK_/ZBS_ text form and every recipient form of spec/addresses.md."""
import hashlib
import re
import struct
from dataclasses import dataclass
from typing import Optional, Tuple

from .encoding import (RIPPLE_ALPHABET, base32_decode, base32_encode, base58_decode, base58check_decode,
                       bech32_decode_plain, segwit_decode, ss58_decode)

ZOOBC, BITCOIN, EMPTY, ESTONIA_EID, ETHEREUM = 0, 1, 2, 3, 4
BITCOIN_P2PKH, BITCOIN_P2SH, BITCOIN_P2WPKH, BITCOIN_P2WSH, BITCOIN_TAPROOT = 5, 6, 7, 8, 9
DATASET, SOLANA, POLKADOT, CARDANO, RIPPLE, TRON, TEZOS = 10, 11, 12, 13, 14, 15, 16

TYPE_NAMES = {0: "ZooBC", 1: "Bitcoin", 3: "Estonia eID", 4: "Ethereum", 5: "Bitcoin P2PKH", 6: "Bitcoin P2SH",
              7: "Bitcoin P2WPKH", 8: "Bitcoin P2WSH", 9: "Bitcoin Taproot", 10: "DataSet", 11: "Solana",
              12: "Polkadot", 13: "Cardano", 14: "Ripple", 15: "Tron", 16: "Tezos"}
CHAINS = {"zbc": "zbc", "zoobc": "zbc", "btc": "btc", "bitcoin": "btc", "eth": "eth", "ethereum": "eth", "evm": "eth",
          "sol": "sol", "solana": "sol", "dot": "dot", "polkadot": "dot", "substrate": "dot", "ada": "ada", "cardano": "ada",
          "xrp": "xrp", "ripple": "xrp", "trx": "trx", "tron": "trx", "xtz": "xtz", "tezos": "xtz", "zbs": "zbs", "dataset": "zbs"}
_HEX = re.compile(r"^[0-9a-fA-F]*$")


def is_hex(s: str, length: Optional[int] = None) -> bool:
    return bool(_HEX.match(s)) and len(s) % 2 == 0 and (length is None or len(s) == length)


def account_type_name(t: int) -> str:
    return TYPE_NAMES.get(t, "type %d" % t)


def payload_length(t: int) -> int:
    """Payload length by account type, as the node parses an envelope."""
    if t in (1, 4, 5, 6, 7, 14, 15, 16):
        return 20
    return 28 if t == 13 else 32


def typed_address(t: int, payload: bytes) -> bytes:
    return struct.pack("<i", t) + payload


def encode_zbc_address(payload: bytes, prefix: str = "ZBC") -> str:
    """PREFIX_ + base32(payload || SHA3-256(payload || prefix)[:3]) in seven groups of eight."""
    if len(payload) != 32 or len(prefix) != 3:
        raise ValueError("address payload must be 32 bytes and the prefix 3 characters")
    check = hashlib.sha3_256(payload + prefix.encode()).digest()[:3]
    b32 = base32_encode(payload + check)
    return prefix + "".join("_" + b32[8 * i:8 * i + 8] for i in range(7))


def decode_zbc_address(text: str) -> Optional[Tuple[str, bytes]]:
    """(upper-case prefix, 32-byte payload) of PREFIX_... (separators _ or -, any case), or None."""
    norm = text.upper()
    if len(norm) < 4 or norm[3] not in "_-":
        return None
    prefix, body = norm[:3], norm[4:].replace("_", "").replace("-", "")
    if len(body) != 56:
        return None
    try:
        raw = base32_decode(body)
    except (ValueError, TypeError):
        return None
    if len(raw) != 35:
        return None
    payload = raw[:32]
    if hashlib.sha3_256(payload + prefix.encode()).digest()[:3] != raw[32:]:
        return None
    return prefix, payload


@dataclass
class ParsedAddress:
    type: int
    payload: bytes
    display: str

    @property
    def bytes(self) -> bytes:
        return typed_address(self.type, self.payload)

    @property
    def type_name(self) -> str:
        return account_type_name(self.type)


def _zbc_form(a: str) -> ParsedAddress:
    d = decode_zbc_address(a)
    if d is None:
        raise ValueError("invalid ZooBC address checksum")
    return ParsedAddress(DATASET if d[0] == "ZBS" else ZOOBC, d[1], a)


def _hinted(a: str, hint: str) -> ParsedAddress:
    if hint == "eth":
        h = a[2:] if a[:2] in ("0x", "0X") else a
        if not is_hex(h, 40):
            raise ValueError("not a 20-byte Ethereum address")
        return ParsedAddress(ETHEREUM, bytes.fromhex(h), a)
    if hint == "sol":
        d = base58_decode(a)
        if d is None or len(d) != 32:
            raise ValueError("not a 32-byte Solana address")
        return ParsedAddress(SOLANA, d, a)
    if hint == "dot":
        ss = ss58_decode(a)
        if ss is None:
            raise ValueError("not a valid SS58 address")
        return ParsedAddress(POLKADOT, ss[1], a)
    if hint in ("zbc", "zbs"):
        return _zbc_form(a)
    return _auto(a)


def _auto(a: str) -> ParsedAddress:
    if len(a) == 42 and a[:2] in ("0x", "0X") and is_hex(a[2:], 40):
        return ParsedAddress(ETHEREUM, bytes.fromhex(a[2:]), a)
    if len(a) > 4 and a[3] in "_-":
        return _zbc_form(a)
    low5 = a[:5].lower()
    if low5.startswith(("bc1", "tb1", "bcrt1")):
        d = segwit_decode(a)
        if d is None:
            raise ValueError("invalid Bitcoin bech32 address")
        _, version, prog = d
        if version == 0 and len(prog) == 20:
            return ParsedAddress(BITCOIN_P2WPKH, prog, a)
        if version == 0 and len(prog) == 32:
            return ParsedAddress(BITCOIN_P2WSH, prog, a)
        if version == 1 and len(prog) == 32:
            return ParsedAddress(BITCOIN_TAPROOT, prog, a)
        raise ValueError("unsupported Bitcoin witness program")
    if a[0] in "13" and 26 <= len(a) <= 35:
        raw = base58_decode(a)
        if raw is not None and len(raw) == 25:
            body = base58check_decode(a)
            if body is not None and len(body) == 21:
                if body[0] == 0x00:
                    return ParsedAddress(BITCOIN_P2PKH, body[1:], a)
                if body[0] == 0x05:
                    return ParsedAddress(BITCOIN_P2SH, body[1:], a)
    if len(a) > 5 and a[:5].lower() == "addr1":
        d = bech32_decode_plain(a)
        if d is not None and d[0] == "addr" and len(d[1]) == 29 and d[1][0] == 0x61:
            return ParsedAddress(CARDANO, d[1][1:], a)
        raise ValueError("invalid Cardano address (expected a mainnet enterprise addr1… address)")
    if a[0] == "T" and len(a) == 34:
        body = base58check_decode(a)
        if body is not None and len(body) == 21 and body[0] == 0x41:
            return ParsedAddress(TRON, body[1:], a)
        raise ValueError("invalid Tron address")
    if a[0] == "r" and 25 <= len(a) <= 35:
        body = base58check_decode(a, RIPPLE_ALPHABET)
        if body is not None and len(body) == 21 and body[0] == 0x00:
            return ParsedAddress(RIPPLE, body[1:], a)
        raise ValueError("invalid Ripple address")
    if a.startswith("tz1"):
        body = base58check_decode(a)
        if body is not None and len(body) == 23 and body[:3] == b"\x06\xa1\x9f":
            return ParsedAddress(TEZOS, body[3:], a)
        raise ValueError("invalid Tezos address")
    ss = ss58_decode(a)
    if ss is not None and len(ss[1]) == 32:
        return ParsedAddress(POLKADOT, ss[1], a)
    if 32 <= len(a) <= 44:
        d = base58_decode(a)
        if d is not None and len(d) == 32:
            return ParsedAddress(SOLANA, d, a)
    if is_hex(a, 64):
        key = bytes.fromhex(a)
        return ParsedAddress(ZOOBC, key, encode_zbc_address(key, "ZBC"))
    raise ValueError("unrecognised address. Supported: ZooBC (ZBC_/ZBS_), Bitcoin, Ethereum, Solana, Polkadot, Cardano, Ripple, Tron, Tezos")


def parse_address(text: str, chain: str = "") -> ParsedAddress:
    """Read a recipient in the order of spec/addresses.md section 3; `chain` forces one reading (--chain)."""
    a = text.strip()
    if not a:
        raise ValueError("empty address")
    if chain:
        hint = CHAINS.get(chain.lower())
        if hint is None:
            raise ValueError("unknown chain %s" % chain)
        try:
            return _hinted(a, hint)
        except ValueError:
            up = a.upper()
            plain = ((len(a) == 42 and a[:2] in ("0x", "0X")) or up.startswith(("ZBC", "ZNK", "ZBS")) or a[0] in "13"
                     or a[:5].lower().startswith(("bc1", "tb1", "bcrt1")) or len(a) == 64)
            if not plain:
                raise
            return _auto(a)
    return _auto(a)


def parse_key32(text: str) -> bytes:
    """A registry key parameter: 64 hex (optionally 0x) or a ZNK_/ZBG_/ZBR_/ZBC_ text address; 32 bytes."""
    if len(text) == 66 and text[3] == "_":
        d = decode_zbc_address(text)
        if d is None:
            raise ValueError("invalid address checksum")
        return d[1]
    h = text[2:] if text[:2] in ("0x", "0X") else text
    if not is_hex(h, 64):
        raise ValueError("key must be a 64-hex string or a ZNK_/ZBG_/ZBR_ address")
    return bytes.fromhex(h)

File: rust/zbc/src/address.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci

//! Account addresses: the ZBC_/ZNK_/ZBS_ text form and every recipient form of spec/addresses.md.

use crate::encoding::*;
use sha3::{Digest, Sha3_256};

pub const ZOOBC: i32 = 0;
pub const BITCOIN: i32 = 1;
pub const EMPTY: i32 = 2;
pub const ESTONIA_EID: i32 = 3;
pub const ETHEREUM: i32 = 4;
pub const BITCOIN_P2PKH: i32 = 5;
pub const BITCOIN_P2SH: i32 = 6;
pub const BITCOIN_P2WPKH: i32 = 7;
pub const BITCOIN_P2WSH: i32 = 8;
pub const BITCOIN_TAPROOT: i32 = 9;
pub const DATASET: i32 = 10;
pub const SOLANA: i32 = 11;
pub const POLKADOT: i32 = 12;
pub const CARDANO: i32 = 13;
pub const RIPPLE: i32 = 14;
pub const TRON: i32 = 15;
pub const TEZOS: i32 = 16;

pub fn account_type_name(t: i32) -> String {
    match t {
        0 => "ZooBC",
        1 => "Bitcoin",
        3 => "Estonia eID",
        4 => "Ethereum",
        5 => "Bitcoin P2PKH",
        6 => "Bitcoin P2SH",
        7 => "Bitcoin P2WPKH",
        8 => "Bitcoin P2WSH",
        9 => "Bitcoin Taproot",
        10 => "DataSet",
        11 => "Solana",
        12 => "Polkadot",
        13 => "Cardano",
        14 => "Ripple",
        15 => "Tron",
        16 => "Tezos",
        _ => return format!("type {t}"),
    }
    .to_string()
}

/// The 4-byte little-endian type followed by the payload.
pub fn typed_address(t: i32, payload: &[u8]) -> Vec<u8> {
    let mut out = t.to_le_bytes().to_vec();
    out.extend_from_slice(payload);
    out
}

/// PREFIX_ + base32(payload || SHA3-256(payload || prefix)[..3]) in seven groups of eight.
pub fn encode_zbc_address(payload: &[u8], prefix: &str) -> Result<String, String> {
    if payload.len() != 32 || prefix.len() != 3 {
        return Err("address payload must be 32 bytes and the prefix 3 characters".into());
    }
    let mut h = Sha3_256::new();
    h.update(payload);
    h.update(prefix.as_bytes());
    let sum = h.finalize();
    let mut raw = payload.to_vec();
    raw.extend_from_slice(&sum[..3]);
    let s = base32_encode(&raw);
    let mut out = prefix.to_string();
    for i in 0..7 {
        out.push('_');
        out.push_str(&s[8 * i..8 * i + 8]);
    }
    Ok(out)
}

/// (upper-case prefix, 32-byte payload) of PREFIX_... (separators _ or -, any case), or None.
/// The 59 significant characters of a ZooBC address: separators (_ -) and whitespace dropped, upper case (addresses.md 2).
pub fn zbc_significant(text: &str) -> String {
    text.chars().filter(|c| *c != '_' && *c != '-' && !c.is_whitespace()).collect::<String>().to_uppercase()
}

/// (upper-case prefix, 32-byte payload) of a ZooBC address in any spelling, or None.
pub fn decode_zbc_address(text: &str) -> Option<(String, Vec<u8>)> {
    let norm = zbc_significant(text);
    if norm.len() < 3 || !norm.is_ascii() {
        return None;
    }
    let prefix = &norm[..3];
    let body = &norm[3..];
    if body.len() != 56 {
        return None;
    }
    let raw = base32_decode(&body)?;
    if raw.len() != 35 {
        return None;
    }
    let mut h = Sha3_256::new();
    h.update(&raw[..32]);
    h.update(prefix.as_bytes());
    let sum = h.finalize();
    if sum[..3] != raw[32..] {
        return None;
    }
    Some((prefix.to_string(), raw[..32].to_vec()))
}

/// A recipient as the envelope carries it.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedAddress {
    pub account_type: i32,
    pub payload: Vec<u8>,
    pub display: String,
}

impl ParsedAddress {
    /// The typed bytes: 4-byte type LE then the payload (36 bytes for a ZooBC account).
    pub fn bytes(&self) -> Vec<u8> {
        typed_address(self.account_type, &self.payload)
    }
    pub fn type_name(&self) -> String {
        account_type_name(self.account_type)
    }
}

fn chain_of(name: &str) -> Option<&'static str> {
    Some(match name.to_lowercase().as_str() {
        "zbc" | "zoobc" => "zbc",
        "btc" | "bitcoin" => "btc",
        "eth" | "ethereum" | "evm" => "eth",
        "sol" | "solana" => "sol",
        "dot" | "polkadot" | "substrate" => "dot",
        "ada" | "cardano" => "ada",
        "xrp" | "ripple" => "xrp",
        "trx" | "tron" => "trx",
        "xtz" | "tezos" => "xtz",
        "zbs" | "dataset" => "zbs",
        _ => return None,
    })
}

/// Shape only: PREFIX then a separator, or the bare form: 59 significant characters, ZBC/ZBS prefix, base32 body.
fn looks_zbc(a: &str) -> bool {
    let b = a.as_bytes();
    if a.len() > 4 && (b[3] == b'_' || b[3] == b'-') {
        return true;
    }
    let n = zbc_significant(a);
    n.len() == 59
        && n.is_ascii()
        && (n.starts_with("ZBC") || n.starts_with("ZBS"))
        && n[3..].bytes().all(|c| b"ABCDEFGHIJKLMNOPQRSTUVWXYZ234567".contains(&c))
}

fn zbc_form(a: &str) -> Result<ParsedAddress, String> {
    let (prefix, payload) = decode_zbc_address(a).ok_or("invalid ZooBC address checksum")?;
    let t = if prefix == "ZBS" { DATASET } else { ZOOBC };
    Ok(ParsedAddress { account_type: t, payload, display: a.to_string() })
}

fn parse_hinted(a: &str, hint: &str) -> Result<ParsedAddress, String> {
    match hint {
        "eth" => {
            let h = a.strip_prefix("0x").or_else(|| a.strip_prefix("0X")).unwrap_or(a);
            if !is_hex(h, 40) {
                return Err("not a 20-byte Ethereum address".into());
            }
            Ok(ParsedAddress { account_type: ETHEREUM, payload: hex::decode(h).unwrap(), display: a.to_string() })
        }
        "sol" => match base58_decode(a, BITCOIN_ALPHABET) {
            Some(d) if d.len() == 32 => Ok(ParsedAddress { account_type: SOLANA, payload: d, display: a.to_string() }),
            _ => Err("not a 32-byte Solana address".into()),
        },
        "dot" => match ss58_decode(a) {
            Some((_, id)) => Ok(ParsedAddress { account_type: POLKADOT, payload: id, display: a.to_string() }),
            None => Err("not a valid SS58 address".into()),
        },
        "zbc" | "zbs" => zbc_form(a),
        _ => parse_auto(a),
    }
}

fn parse_auto(a: &str) -> Result<ParsedAddress, String> {
    let b = a.as_bytes();
    if a.len() == 42 && (a.starts_with("0x") || a.starts_with("0X")) && is_hex(&a[2..], 40) {
        return Ok(ParsedAddress { account_type: ETHEREUM, payload: hex::decode(&a[2..]).unwrap(), display: a.to_string() });
    }
    if looks_zbc(a) {
        return zbc_form(a);
    }
    let low5: String = a.chars().take(5).collect::<String>().to_lowercase();
    if low5.starts_with("bc1") || low5.starts_with("tb1") || low5.starts_with("bcrt1") {
        let (_, version, prog) = segwit_decode(a).ok_or("invalid Bitcoin bech32 address")?;
        return match (version, prog.len()) {
            (0, 20) => Ok(ParsedAddress { account_type: BITCOIN_P2WPKH, payload: prog, display: a.to_string() }),
            (0, 32) => Ok(ParsedAddress { account_type: BITCOIN_P2WSH, payload: prog, display: a.to_string() }),
            (1, 32) => Ok(ParsedAddress { account_type: BITCOIN_TAPROOT, payload: prog, display: a.to_string() }),
            _ => Err("unsupported Bitcoin witness program".into()),
        };
    }
    if (b[0] == b'1' || b[0] == b'3') && (26..=35).contains(&a.len()) {
        if let Some(raw) = base58_decode(a, BITCOIN_ALPHABET) {
            if raw.len() == 25 {
                if let Some(body) = base58check_decode(a, BITCOIN_ALPHABET) {
                    if body.len() == 21 && body[0] == 0x00 {
                        return Ok(ParsedAddress { account_type: BITCOIN_P2PKH, payload: body[1..].to_vec(), display: a.to_string() });
                    }
                    if body.len() == 21 && body[0] == 0x05 {
                        return Ok(ParsedAddress { account_type: BITCOIN_P2SH, payload: body[1..].to_vec(), display: a.to_string() });
                    }
                }
            }
        }
    }
    if a.len() > 5 && a[..5].to_lowercase() == "addr1" {
        if let Some((hrp, bytes)) = bech32_decode_plain(a) {
            if hrp == "addr" && bytes.len() == 29 && bytes[0] == 0x61 {
                return Ok(ParsedAddress { account_type: CARDANO, payload: bytes[1..].to_vec(), display: a.to_string() });
            }
        }
        return Err("invalid Cardano address (expected a mainnet enterprise addr1… address)".into());
    }
    if b[0] == b'T' && a.len() == 34 {
        if let Some(body) = base58check_decode(a, BITCOIN_ALPHABET) {
            if body.len() == 21 && body[0] == 0x41 {
                return Ok(ParsedAddress { account_type: TRON, payload: body[1..].to_vec(), display: a.to_string() });
            }
        }
        return Err("invalid Tron address".into());
    }
    if b[0] == b'r' && (25..=35).contains(&a.len()) {
        if let Some(body) = base58check_decode(a, RIPPLE_ALPHABET) {
            if body.len() == 21 && body[0] == 0x00 {
                return Ok(ParsedAddress { account_type: RIPPLE, payload: body[1..].to_vec(), display: a.to_string() });
            }
        }
        return Err("invalid Ripple address".into());
    }
    if a.starts_with("tz1") {
        if let Some(body) = base58check_decode(a, BITCOIN_ALPHABET) {
            if body.len() == 23 && body[..3] == [0x06, 0xa1, 0x9f] {
                return Ok(ParsedAddress { account_type: TEZOS, payload: body[3..].to_vec(), display: a.to_string() });
            }
        }
        return Err("invalid Tezos address".into());
    }
    if let Some((_, id)) = ss58_decode(a) {
        if id.len() == 32 {
            return Ok(ParsedAddress { account_type: POLKADOT, payload: id, display: a.to_string() });
        }
    }
    if (32..=44).contains(&a.len()) {
        if let Some(d) = base58_decode(a, BITCOIN_ALPHABET) {
            if d.len() == 32 {
                return Ok(ParsedAddress { account_type: SOLANA, payload: d, display: a.to_string() });
            }
        }
    }
    if is_hex(a, 64) {
        let key = hex::decode(a).unwrap();
        let display = encode_zbc_address(&key, "ZBC")?;
        return Ok(ParsedAddress { account_type: ZOOBC, payload: key, display });
    }
    Err("unrecognised address. Supported: ZooBC (ZBC_/ZBS_), Bitcoin, Ethereum, Solana, Polkadot, Cardano, Ripple, Tron, Tezos".into())
}

/// Read a recipient in the order of spec/addresses.md section 3; `chain` forces one reading (--chain).
pub fn parse_address(text: &str, chain: &str) -> Result<ParsedAddress, String> {
    let a = text.trim();
    if a.is_empty() {
        return Err("empty address".into());
    }
    if !chain.is_empty() {
        let hint = chain_of(chain).ok_or_else(|| format!("unknown chain {chain}"))?;
        return match parse_hinted(a, hint) {
            Ok(p) => Ok(p),
            Err(e) => {
                let up = a.to_uppercase();
                let low = a.to_lowercase();
                let plain = (a.len() == 42 && (a.starts_with("0x") || a.starts_with("0X")))
                    || up.starts_with("ZBC") || up.starts_with("ZNK") || up.starts_with("ZBS")
                    || a.starts_with('1') || a.starts_with('3')
                    || low.starts_with("bc1") || low.starts_with("tb1") || low.starts_with("bcrt1") || a.len() == 64;
                if plain { parse_auto(a) } else { Err(e) }
            }
        };
    }
    parse_auto(a)
}

/// A registry key parameter: 64 hex (optionally 0x) or a ZNK_/ZBG_/ZBR_/ZBC_ text address; 32 bytes.
pub fn parse_key32(text: &str) -> Result<Vec<u8>, String> {
    if text.len() == 66 && text.as_bytes()[3] == b'_' {
        return decode_zbc_address(text).map(|(_, p)| p).ok_or_else(|| "invalid address checksum".into());
    }
    let h = text.strip_prefix("0x").or_else(|| text.strip_prefix("0X")).unwrap_or(text);
    if !is_hex(h, 64) {
        return Err("key must be a 64-hex string or a ZNK_/ZBG_/ZBR_ address".into());
    }
    Ok(hex::decode(h).unwrap())
}
